module/zfs/zfs_vnops.c

```diff
@@ -4636,13 +4654,22 @@ zfs_dirty_inode(struct inode *ip, int flags)
 	}
 #endif
 
+top:
 	tx = dmu_tx_create(zfsvfs->z_os);
 
 	dmu_tx_hold_sa(tx, zp->z_sa_hdl, B_FALSE);
 	zfs_sa_upgrade_txholds(tx, zp);
 
-	error = dmu_tx_assign(tx, TXG_WAIT);
+	boolean_t waited = B_FALSE;
+	error = dmu_tx_assign(tx,
+	    waited ? (TXG_NOTHROTTLE | TXG_WAIT) : TXG_NOWAIT);
```

Comment:
I think itit would be more idiomatic, as well as easier to understand, if we do the same thing as other similar code, i.e.:
```
	error = dmu_tx_assign(tx, (waited ? TXG_NOTHROTTLE : 0) | TXG_NOWAIT);
	if (error) {
		if (error == ERESTART) {
			waited = B_TRUE;
			dmu_tx_wait(tx);
			dmu_tx_abort(tx);
			goto top;
		}
		dmu_tx_abort(tx);
		ZFS_EXIT(zfsvfs);
		return (error);
	}
```
If you search for NOTHROTTLE in this file, you'll see 8 other cases that do it this way.

Reply:
There's an extra complication which isn't obvious when reading this code in isolation which is why it was written a little differently.  Despite the fact that this function allows an error to be returned, it's called from `zpl_dirty_inode()` which is a Linux VFS callback functions (.dirty_inode) which must always succeed.

```
        void (*dirty_inode) (struct inode *, int flags);
```

This is why the existing code used `TXG_WAIT` since it is handled slightly differently in `dmu_tx_try_assign()`.

```
       if (spa_suspended(spa)) {
                ...
                if (spa_get_failmode(spa) == ZIO_FAILURE_MODE_CONTINUE &&
                    !(txg_how & TXG_WAIT))
                        return (SET_ERROR(EIO));

                return (SET_ERROR(ERESTART));
        }
```

If we don't use `TXG_WAIT` it's possible the time update could be dropped and there's no way to report it.  Perhaps we should instead add a comment explaining why this one is different.

Reply:
In that case, please add a comment explaining why we need to use TXG_WAIT.  And I think we would need to use TXG_WAIT every time, not just if the first NOWAIT call fails.

Reply:
I'm sorry, I still don't understand how this works.  The first time through, we call `dmu_tx_assign(TXG_NOWAIT)`.  If the pool is suspended, it will return EIO (per the code @behlendorf quoted above).  zfs_dirty_inode() will then `goto out` and return EIO.  According to the comment you added, this would be incorrect behavior, because it "must always succeed".

Reply:
You are right, we should always assign  by `TXG_NOTHROTTLE | TXG_WAIT`, then `dmu_tx_assign` can only return `ERESTART`, and we try again and agian until we assign successfully.

Reply:
```
if (tx->tx_dir != NULL && asize != 0) {
		int err = dsl_dir_tempreserve_space(tx->tx_dir, memory,
		    asize, tx->tx_netfree, &tx->tx_tempreserve_cookie, tx);
		if (err != 0)
			return (err);
}
```
 `dmu_tx_try_assign` may return like `EDQUOT` or `ENOSPC` error in this place, this will make `dirty inode` op failed, if I always retry no matter what error returned by `dmu_tx_assign`, this may lead to infinite loop, do you have any good suggestions?

Reply:
If the required semantic is "must always succeed", then your only options are to retry (potentially leading to an infinite loop), or panic (halt operation).

Reply:
Note that if you get ENOSPC or EDQUOT (rather than ERESTART), then retrying will only work if some other process happens to free up space.  We could consider using dmu_tx_mark_netfree() to greatly reduce the chances of ENOSPC/EDQUOT, by allowing it to use half the slop space, like unlink().  But it's still possible to get ENOSPC, so we'll have to decide how to handle it.  See also the comments above dmu_tx_mark_netfree() and dsl_synctask.h:ZFS_SPACE_CHECK_RESERVED.

Reply:
>  required semantic is "must always succeed"...

The required semantics are that the inode must be marked dirty so it will be written latter.  The code happens to take this opportunity to assign a transaction for the newly dirtied inode.   However, the vfs does provide a second `.write_inode` callback where the transaction assignment could be moved too.  It does allow for failures and properly handle them.

From `Documentation/filesystems/vfs.txt` the required semantics are:

```
        void (*dirty_inode) (struct inode *, int flags);
        int (*write_inode) (struct inode *, int);

  dirty_inode: this method is called by the VFS to mark an inode dirty.

  write_inode: this method is called when the VFS needs to write an
        inode to disc.  The second parameter indicates whether the write
        should be synchronous or not, not all filesystems check this flag.
```

Splitting this up in something which has been on the list to investigate, but hasn't been critical since the existing code in practice works well.  Getting this right and tested on all the supported kernels might also be tricky.

For the purposes on this PR why don't we leaved this code unchanged and continue to use `TXG_WAIT`.  This won't change the behavior and resolves the deadlock at hand.  Then in a follow up PR, if @wgqimut is available he can investigate implementing the `.write_inode` callback.  For frequently dirtied inode there's potentially a significant performance win to be had.